docs/internal/contributing/how-to-convert-config-to-per-tenant-limit.md

```diff
@@ -0,0 +1,181 @@
+- all docs are updated
+
+We will use the [Mimir PR #5312](https://github.com/grafana/mimir/pull/5312) as a reference example, in which we moved the query-frontend option `cache_unaligned_requests` from the `frontend` config section to `limits`.
+[PR #4287](https://github.com/grafana/mimir/pull/4287) is also available for reference.
+
+## Deprecating the Existing Config Option
+
+1. Move the config option description JSON object from its previous location to the "limits" block of `cmd/mimir/config-descriptor.json`.
+
+2. Locate the actual config option in code and mark it as deprecated by renaming the variable and marking it as `hidden` and `deprecated`.
+
+   In the example PR, the existing config option was a field of the `querymiddleware.Config` struct, declared in `pkg/frontend/querymiddleware/roundtrip.go`.
+
+   The line
+
+   ```go
+   CacheUnalignedRequests bool `yaml:"cache_unaligned_requests" category:"advanced"`
+   ```
+
+   was updated to
+
+   ```go
+   DeprecatedCacheUnalignedRequests bool `yaml:"cache_unaligned_requests" category:"advanced" doc:"hidden"` // Deprecated: Deprecated in Mimir 2.10.0, remove in Mimir 2.12.0 (https://github.com/grafana/mimir/issues/5253)
+   ```
+
+3. Remove the binding of the CLI flag from the deprecated config option.
+   The same flag will be bound to the new config option in the next step.
+
+   Binding CLI flags to the config option occurs in the respective config struct's `RegisterFlags` method.
+   In the example PR, the following line was removed:
+
+   ```go
+   f.BoolVar(&cfg.DeprecatedCacheUnalignedRequests, "query-frontend.cache-unaligned-requests", false, "Cache requests that are not step-aligned.")
+   ```
+
+## Adding the New Config Option
+
+1. Add the new config option to the `validation.Limits` struct in `pkg/util/validation/limits.go`.
+
+   Add the option alongside any related `limits` options for the same service component.
+   The new struct field may need a more descriptive name as is now a member of a more generalized config section.
+
+   In the example, the previous field name under `querymiddleware.Config` was `CacheUnalignedRequests`.
+   The new field is named `ResultsCacheForUnalignedQueryEnabled` to align with the naming of related options:
+
+   ```go
+   ResultsCacheForUnalignedQueryEnabled bool `yaml:"cache_unaligned_requests" json:"cache_unaligned_requests" category:"advanced"`
+   ```
+
+2. Add the binding of the CLI flag to the Limits config.
+
+   Again, group it with other related options.
+   Ensure the same default value is maintained.
+
+   In `validation.Limits` struct's `RegisterFlags` method:
+
+   ```go
+   f.BoolVar(&l.ResultsCacheForUnalignedQueryEnabled, "query-frontend.cache-unaligned-requests", false, "Cache requests that are not step-aligned.")
+   ```
+
+3. Expose the tenant-specific overrides for the new config option
+
+   First add the tenant-specific method and docstring to the _interface_ which uses the config option.
```

Comment:
Using an interface like this is specific to the way limits are used in this particular area of the code. Can you note that this pattern of adding the setting to an interface may not apply in all cases?